### DISCUSSION TITLE
How to watch subscribers from icebox?

### DISCUSSION THREAD

**Discussion by yinghaoyu** [zeroc-ice/ice#2203] [Q&A]
Using vim to open the file named `subscribers` with garbled characters.

**Comment by pepone** [comment]
Hi @yinghaoyu 

Can you please tell us the path and the repository of the file?

**Reply by yinghaoyu** [reply]
$ cat config.service
``` bash
#
# This property is used by the administrative client to connect to IceStorm.
#
#IceStorm.TopicManager.Proxy=MonitorIceStorm/TopicManager:default -p 9999

Ice.Default.EncodingVersion=1.0

# This property defines the endpoints on which the IceStorm
# TopicManager listens.
#
IceStorm.TopicManager.Endpoints=default -p 9999

#
# The IceStorm service instance name.
#
IceStorm.InstanceName=MonitorIceStorm

#
# This property defines the endpoints on which the topic
# publisher objects listen. If you want to federate
# IceStorm instances this must run on a fixed port (or use
# IceGrid).
#
IceStorm.Publish.Endpoints=tcp -p 10001:udp -p 10001

#
# TopicManager Tracing
#
# 0 = no tracing
# 1 = trace topic creation, subscription, unsubscription
# 2 = like 1, but with more detailed subscription information
#
IceStorm.Trace.TopicManager=2

#
# Topic Tracing
#
# 0 = no tracing
# 1 = trace unsubscription diagnostics
#
IceStorm.Trace.Topic=1

#
# Subscriber Tracing
#
# 0 = no tracing
# 1 = subscriber diagnostics (subscription, unsubscription, event
#     propagation failures)
#
IceStorm.Trace.Subscriber=1

#
# Flush Tracing (for batch mode transfer flushing)
#
# 0 = no tracing
# 1 = trace activity of flusher thread
#
#IceStorm.Trace.Flush=1

#
# Amount of time in milliseconds between flushes for batch mode
# transfer. The minimum allowable value is 100ms.
#
IceStorm.Flush.Timeout=2000

#
# Network Tracing
#
# 0 = no network tracing
# 1 = trace connection establishment and closure
# 2 = like 1, but more detailed
# 3 = like 2, but also trace data transfer
#
Ice.Trace.Network=3
Ice.Trace.ThreadPool=1

#
# This property defines the home directory of the Freeze
# database environment for the IceStorm service.
#
Freeze.DbEnv.IceStorm.DbHome=db
```
` Freeze.DbEnv.IceStorm.DbHome=db ` specifies the db path, and the file named ` subscribers ` is  in this db directory.

**Reply by pepone** [reply]
Files in the db directory are binary files, Freeze uses Oracle Berkeley DB as the underlying database.

**Reply by yinghaoyu** [reply]
How can I get all the subscriber information for this db file? Are there any examples for reference？

**Reply by yinghaoyu** [reply]
Or is there another way to see the subject to which the subscriber subscribed and the subscriber's ip？

**Reply by pepone** [reply]
Can you clarify what Ice version you are using?

**Reply by yinghaoyu** [reply]
`Ice 3.4.2`

**Reply by yinghaoyu** [reply]
Hi

**Reply by pepone** [reply]
Sorry for the delay.

You can use the `icestormadmin` command to list the subscribers for each topic.

See https://doc.zeroc.com/ice/3.7/ice-services/icestorm/icestorm-administration

You can do the same programmatically using the TopicManager and Topic interfaces

https://github.com/zeroc-ice/ice/blob/5c68ccc3d38daf890b4f7390d05ac14f6629e937/slice/IceStorm/IceStorm.ice#L85

Note 3.4.2 is out of support, you should consider upgrading to 3.7 release.

Cheers,
Jose

**Reply by yinghaoyu** [reply]
Thanks for your reply.
Sorry for the version I had mentioned above. `3.6.2` is the correct version being used. But it seems not working by using `icestormadmin -e topics`. It shows as below,
```bash
!! 06/12/24 16:42:12.096 icestormadmin: error: IPEndpointI.cpp:429: Ice::EndpointParseException:error while parsing endpoint `no argument provided for -p option in endpoint `tcp  -p -t 3000''
```

**Reply by pepone** [reply]
As the error says you are using an invalid `-p` option in your endpoint. `-p` requires and argument see https://doc.zeroc.com/ice/3.7/client-side-features/proxies/proxy-and-endpoint-syntax

**Reply by yinghaoyu** [reply]
> $ cat config.service
> 
> ```shell
> #
> # This property is used by the administrative client to connect to IceStorm.
> #
> #IceStorm.TopicManager.Proxy=MonitorIceStorm/TopicManager:default -p 9999
> 
> Ice.Default.EncodingVersion=1.0
> 
> # This property defines the endpoints on which the IceStorm
> # TopicManager listens.
> #
> IceStorm.TopicManager.Endpoints=default -p 9999
> 
> #
> # The IceStorm service instance name.
> #
> IceStorm.InstanceName=MonitorIceStorm
> 
> #
> # This property defines the endpoints on which the topic
> # publisher objects listen. If you want to federate
> # IceStorm instances this must run on a fixed port (or use
> # IceGrid).
> #
> IceStorm.Publish.Endpoints=tcp -p 10001:udp -p 10001
> 
> #
> # TopicManager Tracing
> #
> # 0 = no tracing
> # 1 = trace topic creation, subscription, unsubscription
> # 2 = like 1, but with more detailed subscription information
> #
> IceStorm.Trace.TopicManager=2
> 
> #
> # Topic Tracing
> #
> # 0 = no tracing
> # 1 = trace unsubscription diagnostics
> #
> IceStorm.Trace.Topic=1
> 
> #
> # Subscriber Tracing
> #
> # 0 = no tracing
> # 1 = subscriber diagnostics (subscription, unsubscription, event
> #     propagation failures)
> #
> IceStorm.Trace.Subscriber=1
> 
> #
> # Flush Tracing (for batch mode transfer flushing)
> #
> # 0 = no tracing
> # 1 = trace activity of flusher thread
> #
> #IceStorm.Trace.Flush=1
> 
> #
> # Amount of time in milliseconds between flushes for batch mode
> # transfer. The minimum allowable value is 100ms.
> #
> IceStorm.Flush.Timeout=2000
> 
> #
> # Network Tracing
> #
> # 0 = no network tracing
> # 1 = trace connection establishment and closure
> # 2 = like 1, but more detailed
> # 3 = like 2, but also trace data transfer
> #
> Ice.Trace.Network=3
> Ice.Trace.ThreadPool=1
> 
> #
> # This property defines the home directory of the Freeze
> # database environment for the IceStorm service.
> #
> Freeze.DbEnv.IceStorm.DbHome=db
> ```
> 
> `Freeze.DbEnv.IceStorm.DbHome=db` specifies the db path, and the file named `subscribers` is in this db directory.

As you can see in this file, `IceStorm.TopicManager.Endpoints` had already been set. I have no cure about the errors.

**Reply by pepone** [reply]
icestormadmin does not use `IceStorm.TopicManager.Endpoints`, you probably want to use `IceStormAdmin.TopicManager.Default` see the linked doc for details.
